[AMDGPU] Add MachineVerifier check to detect illegal copies from vector register to SGPR

AditiRM opened the pull request:
Addition of a check in the MachineVerifier to detect and report illegal vector registers to SGPR copies in the AMDGPU backend, ensuring correct code generation.  
We can enforce this check only after SIFixSGPRCopies pass.
This is half-fix in the pipeline with the help of isSSA MachineFuction property, the check is happening for passes after phi-node-elimination.

Commit:
[AMDGPU] Add MachineVerifer check to detect illegal copies from VGPR to SGPR

jayfoad commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -4613,15 +4613,41 @@ static bool isSubRegOf(const SIRegisterInfo &TRI,
+    // FIXME: At this point the COPY verify is done only for non-ssa forms.
+    // Find a better property to recognize the point where instruction selection
+    // is just done.
+    if (!MRI.isSSA() && MI.isCopy())
```
Why is the isSSA check needed?

arsenm replied:
Because we lack the appropriate property to express the real condition, and this is a half measure. We really want a custom IsSelectedAndSGPRCopiesFixed property, otherwise the natural check will fail before SIFixSGPRCopies

jayfoad replied:
OK, then the comment should say explicitly say that we can only enforce this after SIFixSGPRCopies.

cdevadas replied:
I was wondering; the existing ` MachineFunctionProperties::Property::Selected` currently set in the GISel pipeline when selection is completed, should be set in the SelectionDAG as well (mostly after finalize-isel) for consistency. MachineFunctions currently don't have this property set for SelectionDAG.
If this field is rightly set, we could catch the illegal copies soon after finalize-isel, by then SIFixSGPRCopies would have been invoked.

arsenm replied:
The problem is if selectiondag sets it, the natural place it would be set in selection is still before SIFixSGPRCopies

cdevadas replied:
Isn't FinalizeISel that marks the end of selection? If not, the pass name is misleading.

arsenm replied:
Yes and no, and I don't think it's really misleading. It finalizes the already selected function into a verifiable form

cdevadas replied:
Improve the comment to include Jay's suggestion.
"explicitly say that we can only enforce this after SIFixSGPRCopies."

AditiRM replied:
Improved the comment.

jayfoad commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -4613,15 +4613,41 @@ static bool isSubRegOf(const SIRegisterInfo &TRI,
          SubReg.getReg() == SuperVec.getReg();
 }
 
+// Verify the illgal copy from VGPR to SGPR for generic opcode COPY
```
Typo "illegal"

jayfoad commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -4613,15 +4613,41 @@ static bool isSubRegOf(const SIRegisterInfo &TRI,
+  const MachineOperand &Dst = MI.getOperand(0);
+  const MachineOperand &Src = MI.getOperand(1);
+
+  if (Dst.isReg() && Src.isReg()) {
```
MachineVerifier already enforces that both operands are registers so you should not need to check here.

AditiRM replied:
Done

jayfoad commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -4613,15 +4613,41 @@ static bool isSubRegOf(const SIRegisterInfo &TRI,
+    Register DstReg = Dst.getReg();
+    Register SrcReg = Src.getReg();
+    // This is a check for copy from an VGPR to SGPR
+    if (RI.isVGPR(MRI, SrcReg) && RI.isSGPRReg(MRI, DstReg)) {
```
Does this also diagnose AGPR to SGPR copies?

cdevadas replied:
Also, this only validates in one direction. shouldn't we check the other way as well?
SGPR -> VectorClass
VectorClass -> SGPR

jayfoad replied:
No! COPY from SGPR to VGPR is fine.

AditiRM replied:
Added the illegal copy check from AGPR to SGPR.

arsenm commented on llvm/test/CodeGen/AMDGPU/fix-illegal-vgpr-copies.mir:
```diff
@@ -0,0 +1,29 @@
+# RUN: not --crash llc -march=amdgcn -mcpu=gfx1200 -run-pass=machineverifier %s 2>&1 | FileCheck -check-prefix=ERR %s
```
This belongs in test/MachineVerifier. Also you can just use -run-pass=none, the verifier runs on the input anyway

AditiRM replied:
Done.

arsenm commented on llvm/test/CodeGen/AMDGPU/fix-illegal-vgpr-copies.mir:
```diff
@@ -0,0 +1,29 @@
+    ; ERR: *** Bad machine code: illegal copy from VGPR to SGPR ***
+    ; ERR: instruction: $sgpr1 = COPY $vgpr0
+    $sgpr1 = COPY $vgpr0
+
```
Test the AGPR case, and test some register tuple classes.

AditiRM replied:
Updated the testcase with AGPR and tuple cases

arsenm commented on llvm/test/CodeGen/AMDGPU/phi-moveimm-subreg-input.mir:
```diff
@@ -0,0 +1,30 @@
+# RUN: not --crash llc  -mtriple=amdgcn -mcpu=gfx900 -verify-machineinstrs -run-pass=si-fix-sgpr-copies -o - %s 2>&1 | FileCheck -check-prefix=GCN %s
+
```
This looks like an unrelated test?

AditiRM replied:
In this specific testcase(phi-vgpr-input-moveimm.mir) it is throwing illegal copy error as expected, and it is because isSSA considers subreg defs invalid in SSA. So I have extracted the specific subtest & converted it into not crash. Please let me know if this is valid.

Reference : https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/MIRParser/MIRParser.cpp#L361

AditiRM replied:
@arsenm Please take a look at this testcase change.

cdevadas replied:
I guess this test should be removed. See my comment below in the original test.

AditiRM replied:
I removed this testcase instance.

Commit:
Handle review comments

Addition of illegal AGPR to SGPR copies and update with review changes

cdevadas commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -4613,15 +4613,45 @@ static bool isSubRegOf(const SIRegisterInfo &TRI,
+bool SIInstrInfo::verifyCopy(const MachineInstr &MI,
+                             const MachineRegisterInfo &MRI,
+                             StringRef &ErrInfo) const {
+  const MachineOperand &Dst = MI.getOperand(0);
+  const MachineOperand &Src = MI.getOperand(1);
+
+  Register DstReg = Dst.getReg();
+  Register SrcReg = Src.getReg();
```
```suggestion
  Register DstReg = MI.getOperand(0).getReg();
  Register SrcReg = MI.getOperand(1).getReg();
```

AditiRM replied:
Done

cdevadas commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -4613,15 +4613,45 @@ static bool isSubRegOf(const SIRegisterInfo &TRI,
+  if (RI.isAGPR(MRI, SrcReg) && RI.isSGPRReg(MRI, DstReg)) {
+    ErrInfo = "illegal copy from AGPR to SGPR";
+    return false;
+  }
```
I would prefer a single check with RI.isVectorRegister() that includes both VGPRs and AGPRs. Change the error message to "illegal copy from vector register to SGPR"

AditiRM replied:
Updated the error message with single check.

cdevadas commented on llvm/test/CodeGen/AMDGPU/phi-vgpr-input-moveimm.mir:
```diff
@@ -32,38 +32,6 @@ body:             |
 ...
 
----
-# GCN-LABEL: name: phi_moveimm_subreg_input
```
This test doesn't look legitimate to be invoked by a pass that needs SSA from. The presence of the subreg def makes this test a non-SSA one.

AditiRM replied:
So should I remove this testcase itself ? Also do we need to add this extracted testcase again to the original one ?

cdevadas replied:
Yes, avoid this new test case. I recommend deleting this instance from the original test file.

AditiRM replied:
Done

cdevadas commented:
Change the PR heading and the description to include vector register instead of VGPR.

Commit:
Handle Review comments

introduced vector register illegal copies for both VGPR and AGPR, removed subreg-input testcase scenario, updated the testcase error checks accordingly

cdevadas reviewed (approved):
LGTM

arsenm commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -4613,15 +4613,36 @@ static bool isSubRegOf(const SIRegisterInfo &TRI,
   const MachineFunction *MF = MI.getParent()->getParent();
   const MachineRegisterInfo &MRI = MF->getRegInfo();
 
+  if (SIInstrInfo::isGenericOpcode(MI.getOpcode())) {
```
The isGenericOpcode check is redundant with the isCopy check

AditiRM replied:
Yes, fixed.

arsenm commented on llvm/test/MachineVerifier/fix-illegal-copies.mir:
```diff
@@ -0,0 +1,60 @@
+# RUN: not --crash llc -march=amdgcn -mcpu=gfx1200 -run-pass=none -o /dev/null %s 2>&1 | FileCheck %s    
+# REQUIRES: amdgpu-registered-target    
```
Test file name doesn't indicate it's an AMDGPU test. We probably should have an AMDGPU directory at this point too

AditiRM replied:
Implemented this in new PR and updated in latest commit.

arsenm commented on llvm/test/CodeGen/AMDGPU/phi-vgpr-input-moveimm.mir:
```diff
@@ -32,38 +32,6 @@ body:             |
-
-  bb.3:
-    successors: %bb.2
-    undef %1.sub0:sreg_64 = COPY %0
```
Can you move the sub register indices to the phi input operands, instead of deleting the test

arsenm replied:
ping?

AditiRM replied:
Updated the testcase. Please check @arsenm.

Commit:
Merge branch 'main' into machine-verify

Commit:
Handle review comments

Removed the redundancy of isGenericOpcode check, now isCopy check is indepenent of Generic Opcodes (GISEL pipeline) & SDAG piepeline

Commit:
Merge branch 'main' into machine-verify

Commit:
Move testcase into AMDGPU directory

Commit:
Modify subreg testcase

Commit:
Fix trailing whitespace in SIInstrInfo.cpp

Commit:
Format testcase

arsenm commented on llvm/test/CodeGen/AMDGPU/phi-vgpr-input-moveimm.mir:
```diff
@@ -34,36 +34,35 @@ body:             |
   bb.2:
     successors: %bb.3
-    %3:sreg_64 = PHI %1, %bb.3, %2, %bb.1
+    %3:sreg_32 = PHI %1, %bb.3, %2, %bb.1
```
%1 should probably be a full copy, and this operand %1.sub0 (same for %2)

AditiRM replied:
Can you please explain why do you think we should be introducing subreg in PHI input itself ?

arsenm replied:
When the phi is being processed, the only relevant sub registers would come from the phi's own operands, not some other instructions

Commit:
Merge branch 'main' into machine-verify

cdevadas commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -4600,15 +4600,37 @@ static bool isSubRegOf(const SIRegisterInfo &TRI,
+  if (!MRI.isSSA() && MI.isCopy())
+    return verifyCopy(MI, MRI, ErrInfo);
+
+  if (SIInstrInfo::isGenericOpcode(MI.getOpcode())) {
```
No need of braces.

cdevadas commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -4600,15 +4600,37 @@ static bool isSubRegOf(const SIRegisterInfo &TRI,
+  // FIXME: At this point the COPY verify is done only for non-ssa forms.
+  // Find a better property to recognize the point where instruction selection
+  // is just done.
+  // We can only enforce this check after SIFixSGPRCopies pass.
```
The last sentence is incomplete. May be, something like the following:
We can only enforce this check after SIFixSGPRCopies so that the illegal copies are legalized and thereafter we don't expect a pass inserting similar copies.

Commit:
Handle review changes

arsenm commented:
> LLVM Buildbot has detected a new failure on builder `llvm-clang-x86_64-expensive-checks-debian` running on `gribozavr4` while building `llvm` at step 6 "test-build-unified-tree-check-all".
> 
> Full details are available at: https://lab.llvm.org/buildbot/#/builders/16/builds/5632
> 
> Here is the relevant piece of the build log for the reference

Fixed f9bd08382a4ad84c06e0b572d3b7fc3ecdb81898
